ARROW-14708: [C++] Adding missing abseil dependencies to enable static flight build

rok opened the pull request:
[ARROW-14708](https://issues.apache.org/jira/browse/ARROW-14708)

lidavidm commented:
Thanks! Unfortunately I still get a lot of linker errors using the repro from JIRA, but at least the Abseil issues are eliminated. I think the remaining issues are just because of the order of libraries/needing to repeat libraries.

This is what CMake generates, which fails with linker errors:
```
/usr/bin/c++ -O3 -DNDEBUG  CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example  /home/lidavidm/Code/upstream/arrow-14708/install/lib/libarrow.a  /home/lidavidm/Code/upstream/arrow-14708/install/lib/libarrow_flight.a  -pthread  /usr/lib/x86_64-linux-gnu/libssl.so  /usr/lib/x86_64-linux-gnu/libcrypto.so  -lrt  /home/lidavidm/Code/upstream/arrow-14708/install/lib/libarrow_bundled_dependencies.a
```
The errors indicate `libarrow_flight.a` is missing symbols from `libarrow.a` and the bundled gRPC is missing symbols from OpenSSL.

This works:
```
/usr/bin/c++ -O3 -DNDEBUG  CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example  /home/lidavidm/Code/upstream/arrow-14708/install/lib/libarrow_flight.a  /home/lidavidm/Code/upstream/arrow-14708/install/lib/libarrow.a  -pthread /home/lidavidm/Code/upstream/arrow-14708/install/lib/libarrow_bundled_dependencies.a  /usr/lib/x86_64-linux-gnu/libssl.so  /usr/lib/x86_64-linux-gnu/libcrypto.so  -lrt
```
i.e. link libarrow_flight, then libarrow, then libarrow_bundled_dependencies, then OpenSSL.

Flipping the order of libraries in the CMakeLists.txt fixes the first issue, so I suppose this is user error, or we're missing some dependency expressed in CMake(?).

The installed ArrowFlightTargets.cmake does *not* declare arrow_static in INTERFACE_LINK_LIBRARIES - my CMake understanding is poor but perhaps we need to do so? (It is odd because arrow_static is in the STATIC_LINK_LIBS in flight/CMakeLists.txt.) Also, ArrowTargets.cmake declares OpenSSL as part of arrow_static's INTERFACE_LINK_LIBRARIES, but ArrowConfig.cmake does not do the same for arrow_bundled_dependencies.

lidavidm commented:
This diff gets CMake to repeat `libarrow.a` after `libarrow_flight.a`, so we don't have to order the Arrow libraries ourselves in our "consumer" CMakeLists.txt:

```diff
diff --git a/cpp/src/arrow/flight/CMakeLists.txt b/cpp/src/arrow/flight/CMakeLists.txt
index 8a3228e50..846a420f9 100644
--- a/cpp/src/arrow/flight/CMakeLists.txt
+++ b/cpp/src/arrow/flight/CMakeLists.txt
@@ -170,7 +170,9 @@ add_arrow_lib(arrow_flight
               ${ARROW_FLIGHT_LINK_LIBS}
               STATIC_LINK_LIBS
               arrow_static
-              ${ARROW_FLIGHT_LINK_LIBS})
+              ${ARROW_FLIGHT_LINK_LIBS}
+              STATIC_INSTALL_INTERFACE_LIBS
+              arrow_static)
 
 foreach(LIB_TARGET ${ARROW_FLIGHT_LIBRARIES})
   target_compile_definitions(${LIB_TARGET} PRIVATE ARROW_FLIGHT_EXPORTING)
```

And this additional diff, while questionable, fixes the rest of the link errors:

```diff
diff --git a/cpp/src/arrow/ArrowConfig.cmake.in b/cpp/src/arrow/ArrowConfig.cmake.in
index 6209baeec..750d3aba6 100644
--- a/cpp/src/arrow/ArrowConfig.cmake.in
+++ b/cpp/src/arrow/ArrowConfig.cmake.in
@@ -87,6 +87,10 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
       set_target_properties(
         arrow_static PROPERTIES INTERFACE_LINK_LIBRARIES
         "${arrow_static_interface_link_libraries};arrow_bundled_dependencies")
+      # Propagate dependencies like OpenSSL
+      set_target_properties(
+        arrow_bundled_dependencies PROPERTIES INTERFACE_LINK_LIBRARIES
+        "${arrow_static_interface_link_libraries}")
     endif()
   endif()
 endif()
```

rok commented:
Thanks @lidavidm I was gonna ask for ideas about the flight linking errors! I'll include your diffs and test.

rok commented:
@lidavidm I'm able to compile this locally now. I've also reduced the abseil dependencies (I think to a minimum).
Do we want further review on this?

lidavidm commented:
@github-actions crossbow submit -g nightly

lidavidm commented:
@kou Sorry to trouble you, but would you be able to look over the CMake changes here? I'm not sure if the changes to ArrowConfig.cmake.in are strictly correct.

kou commented on cpp/src/arrow/ArrowConfig.cmake.in:
```diff
@@ -87,6 +87,10 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
+        # Propagate dependencies like OpenSSL
+        set_target_properties(
+          arrow_bundled_dependencies PROPERTIES INTERFACE_LINK_LIBRARIES
+          "${arrow_static_interface_link_libraries}")
```
I don't know why this is needed.
Is there a use-case that uses `arrow_bundled_dependencies` without `arrow_static`? 

BTW, indent for this code is broken.

lidavidm replied:
Ah, the problem here is the order of libraries to the linker. As described above, currently, the order is OpenSSL, then libarrow_bundled_dependencies, so linking will fail because libarrow_bundled_dependencies itself also depends on OpenSSL. Adding this gets CMake to repeat OpenSSL after libarrow_bundled_dependencies and then the linker will find the necessary symbols.

kou replied:
I see. Then we should list OpenSSL in `INTERFACE_LINK_LIBRARIES` of `arrow_bundled_dependencies` not `arrow_static` and `INTERFACE_LINK_LIBRARIES` of `arrow_static` should list only `arrow_bundled_dependencies` something like this:

```diff
diff --git a/cpp/src/arrow/ArrowConfig.cmake.in b/cpp/src/arrow/ArrowConfig.cmake.in
index 6209baeec..ac2b9a027 100644
--- a/cpp/src/arrow/ArrowConfig.cmake.in
+++ b/cpp/src/arrow/ArrowConfig.cmake.in
@@ -38,6 +38,7 @@ set(ARROW_LIBRARY_PATH_SUFFIXES "@ARROW_LIBRARY_PATH_SUFFIXES@")
 set(ARROW_INCLUDE_PATH_SUFFIXES "@ARROW_INCLUDE_PATH_SUFFIXES@")
 set(ARROW_SYSTEM_DEPENDENCIES "@ARROW_SYSTEM_DEPENDENCIES@")
 set(ARROW_BUNDLED_STATIC_LIBS "@ARROW_BUNDLED_STATIC_LIBS@")
+set(ARROW_STATIC_INSTALL_INTERFACE_LIBS "@ARROW_STATIC_INSTALL_INTERFACE_LIBS@")
 
 include("${CMAKE_CURRENT_LIST_DIR}/ArrowOptions.cmake")
 
@@ -79,14 +80,9 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
         PROPERTIES
           IMPORTED_LOCATION
           "${arrow_lib_dir}/${CMAKE_STATIC_LIBRARY_PREFIX}arrow_bundled_dependencies${CMAKE_STATIC_LIBRARY_SUFFIX}"
+          INTERFACE_LINK_LIBRARIES
+          "${ARROW_STATIC_INSTALL_INTERFACE_LIBS}"
         )
-
-      get_property(arrow_static_interface_link_libraries
-                  TARGET arrow_static
-                  PROPERTY INTERFACE_LINK_LIBRARIES)
-      set_target_properties(
-        arrow_static PROPERTIES INTERFACE_LINK_LIBRARIES
-        "${arrow_static_interface_link_libraries};arrow_bundled_dependencies")
     endif()
   endif()
 endif()
diff --git a/cpp/src/arrow/CMakeLists.txt b/cpp/src/arrow/CMakeLists.txt
index 5736c557b..66a24fbfc 100644
--- a/cpp/src/arrow/CMakeLists.txt
+++ b/cpp/src/arrow/CMakeLists.txt
@@ -526,6 +526,9 @@ endif()
 
 if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
   string(APPEND ARROW_PC_LIBS_PRIVATE " -larrow_bundled_dependencies")
+  set(ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL "arrow_bundled_dependencies")
+else()
+  set(ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
 endif()
 # Need -latomic on Raspbian.
 # See also: https://issues.apache.org/jira/browse/ARROW-12860
@@ -557,7 +560,7 @@ add_arrow_lib(arrow
               SHARED_INSTALL_INTERFACE_LIBS
               ${ARROW_SHARED_INSTALL_INTERFACE_LIBS}
               STATIC_INSTALL_INTERFACE_LIBS
-              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
+              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL})
 
 add_dependencies(arrow ${ARROW_LIBRARIES})
 
```

kou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -3476,13 +3476,28 @@ macro(build_grpc)
-      absl::time)
+      absl::throw_delegate
+      absl::time
+      absl::time_zone)
```
Can we create a one-linear(?) to collect this list like https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L2659-L2667 ?

It may be better that we have a shell script that generates the list and include it from this file:

```bash
# cpp/build-support/list-grpc-absl-libraries.sh
...
echo "set(GRPC_GPR_ABSL_LIBRARIES absl::...)"
```

```console
$ cpp/build-support/list-grpc-absl-libraries.sh > cpp/cmake_modules/gRPCVariables.cmake
```

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 66d04acae..36565fdbf 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -3475,14 +3475,7 @@ macro(build_grpc)
                         PROPERTIES IMPORTED_LOCATION "${GRPC_STATIC_LIBRARY_UPB}"
                                    INTERFACE_INCLUDE_DIRECTORIES "${GRPC_INCLUDE_DIR}")
 
-  set(GRPC_GPR_ABSL_LIBRARIES
-      absl::base
-      absl::statusor
-      absl::status
-      absl::cord
-      absl::strings
-      absl::synchronization
-      absl::time)
+  include(gRPCVariables)
   add_library(gRPC::gpr STATIC IMPORTED)
   set_target_properties(gRPC::gpr
                         PROPERTIES IMPORTED_LOCATION "${GRPC_STATIC_LIBRARY_GPR}"
```

kou reviewed (commented):
@lidavidm No problem. :-)

rok commented:
Thanks for the feedback @kou and @lidavidm! I've moved things around as suggested. My cmake skills are lacking.

I've also added a [shell script to extract the `gRPCVariables`](https://github.com/apache/arrow/blob/7d55e3cd32f83588ac12041b61105dc37861f4a3/cpp/build-support/list-absl-libraries.sh). It extracts Abseil and gRPC dependencies at the moment, but I am not sure about it's correctness. Could one of you please take a look and suggest improvements if needed?

rok commented:
I'm getting:
```
CMake Error: install(EXPORT "arrow_targets" ...) includes target "arrow_static" which requires target "arrow_bundled_dependencies" that is not in any export set.
```
When running:
```
cmake -DARROW_DEPENDENCY_SOURCE=BUNDLED ..
```
I'll continue tomorrow but would appreciate pointers :)

lidavidm commented:
Just a thought, maybe the `create_merged_static_lib` call needs to be moved before `add_arrow_lib` so that it exists before we use it?

rok commented:
@kou thanks for the pointers!
I've pushed the change you proposed and hit another error (below). If you can spare time on this I'd be thankful but it's also a nice learning opportunity for me in case you're busy :).

```bash
 -Wall -fno-semantic-interposition -msse4.2 
CMake Error at cmake_modules/BuildUtils.cmake:469 (install):
  install EXPORT given unknown argument "ArrowTargets.cmake".
Call Stack (most recent call first):
  src/arrow/CMakeLists.txt:539 (add_arrow_lib)


-- Creating bundled static library target arrow_bundled_dependencies at /home/rok/Documents/repos/arrow/cpp/build/release/libarrow_bundled_dependencies.a
CMake Error at cmake_modules/BuildUtils.cmake:469 (install):
  install EXPORT given unknown argument "ArrowTestingTargets.cmake".
Call Stack (most recent call first):
  src/arrow/CMakeLists.txt:601 (add_arrow_lib)


CMake Error at cmake_modules/BuildUtils.cmake:469 (install):
  install EXPORT given unknown argument "ArrowDatasetTargets.cmake".
Call Stack (most recent call first):
  src/arrow/dataset/CMakeLists.txt:51 (add_arrow_lib)


-- Found Python3: /home/rok/miniconda3/envs/pyarrow-dev/bin/python3.7 (found version "3.7.12") found components: Interpreter Development NumPy Development.Module Development.Embed 
CMake Error at cmake_modules/BuildUtils.cmake:469 (install):
  install EXPORT given unknown argument "ArrowPythonTargets.cmake".
Call Stack (most recent call first):
  src/arrow/python/CMakeLists.txt:70 (add_arrow_lib)


CMake Error at cmake_modules/BuildUtils.cmake:469 (install):
  install EXPORT given unknown argument "ParquetTargets.cmake".
Call Stack (most recent call first):
  src/parquet/CMakeLists.txt:240 (add_arrow_lib)
```

rok commented:
@github-actions crossbow submit -g nightly

rok commented:
@github-actions crossbow submit -g nightly

rok commented:
@kou could you please review if this change makes sense?

kou commented on cpp/cmake_modules/gRPCVariables.cmake:
```diff
@@ -0,0 +1,25 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Did you generate this file by `list-absl-libraries.sh`?
Could you add a warning comment to this file something like "DO NOT EDIT THIS FILE MANUALLY. YOU SHOULD UPDATE THIS FILE BY list-absl-libraries.sh ..."? (The warning comment should be generated by `list-absl-libraries.sh` automatically.)

rok replied:
Will do!

kou reviewed (commented):
There are many failures in nightly jobs. Could you rebase on mater and re-run nightly jobs to compare with the current nightly jobs result:  https://lists.apache.org/list.html?builds@arrow.apache.org ?

rok commented:
@github-actions crossbow submit -g nightly

rok commented:
@github-actions crossbow submit -g nightly

kszucs commented:
@rok could you please trigger more selective builds with this frequency and trigger the big nightly group once those builds are passing? Additionally I'm working on the release which requires more crossbow bandwidth but the queue is filled now for hours.

rok commented:
@kszucs apologies for taking the resources! I accidentally triggered on the wrong commit and reran quickly on the right one. I'll be more selective in the future. Thanks for the feedback :).

rok commented:
I checked the failing builds that did not fail in the nightly:

Seems unrelated:
* test-conda-python-3.9-spark-master
* test-ubuntu-20.04-cpp-thread-sanitizer
* ubuntu-hirsute-amd64

Seems flaky:
* debian-bookworm-arm64 - GitHub 502 on `archery crossbow`
* homebrew-r-autobrew - Minio issue with R
* python-sdist - GDB issue
* test-build-cpp-fuzz - GitHub permissions issue
* test-fedora-33-python-3 - dependency download issue
* test-r-rstudio-r-base-4.1-opensuse15 - dependency download issue
* wheel-manylinux2014-cp39-arm64 - dependency download issue
* wheel-manylinux2014-cp39-amd64 - vcpkg download issue
* wheel-manylinux2014-cp310-arm64 - vcpkg download issue
* wheel-manylinux2014-cp38-arm64 - vcpkg download issue

ZSTD (undefined reference to `ZSTD_<Something>}’ …):
* example-cpp-minimal-build-static
* example-cpp-minimal-build-static-system-dependency

Unclear:
* ubuntu-bionic-arm64 - errored and did not run

I'll check if ZSTD is related to the change and report back.

kou commented:
@github-actions crossbow submit example-cpp-minimal-build-static*

kou commented:
@github-actions crossbow submit ubuntu-bionic-arm64

rok commented:
@kou I'm still working on the missing symbols, please don't merge :)

rok commented:
@lidavidm using [your repro](https://issues.apache.org/jira/browse/ARROW-14708) this change now only complains of flight not having grpc symbols. Any idea how to approach this?
```
/Library/Developer/CommandLineTools/usr/bin/c++ -O3 -DNDEBUG -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example \
        /Users/rok/builds/arrow-arm64/lib/libarrow.a \
        /Users/rok/builds/arrow-arm64/lib/libarrow_flight.a \
        /opt/homebrew/opt/openssl@3/lib/libssl.a \
        /opt/homebrew/opt/openssl@3/lib/libcrypto.a \
        /Users/rok/builds/arrow-arm64/lib/libarrow_bundled_dependencies.a
```
```
Undefined symbols for architecture arm64:
  "grpc::ServerBuilder::BuildAndStart()", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::ServerBuilder::RegisterService(grpc::Service*)", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::ServerBuilder::AddListeningPort(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<grpc::ServerCredentials>, int*)", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::ServerBuilder::SetOption(std::__1::unique_ptr<grpc::ServerBuilderOption, std::__1::default_delete<grpc::ServerBuilderOption> >)", referenced from:
      grpc::ServerBuilder& grpc::ServerBuilder::AddChannelArgument<int>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int const&) in libarrow_flight.a(server.cc.o)
  "grpc::ServerBuilder::ServerBuilder()", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::ServerBuilder::~ServerBuilder()", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::CreateAuthContext(grpc_call*)", referenced from:
      grpc::ServerContextBase::auth_context() const in libarrow_flight.a(server.cc.o)
  "grpc::ServerContextBase::AddInitialMetadata(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      arrow::flight::(anonymous namespace)::GrpcAddCallHeaders::AddHeader(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libarrow_flight.a(server.cc.o)
  "grpc::ServerContextBase::AddTrailingMetadata(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      arrow::flight::internal::ToGrpcStatus(arrow::Status const&, grpc::ServerContext*) in libarrow_flight.a(internal.cc.o)
  "grpc::SslServerCredentials(grpc::SslServerCredentialsOptions const&)", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::g_core_codegen_interface", referenced from:
      grpc::ServerInterface::Shutdown() in libarrow_flight.a(server.cc.o)
      grpc::CompletionQueue::Pluck(grpc::internal::CompletionQueueTag*) in libarrow_flight.a(server.cc.o)
      grpc::internal::CallOpSet<grpc::internal::CallOpRecvMessage<arrow::flight::protocol::HandshakeRequest>, grpc::internal::CallNoOp<2>, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::FinalizeResult(void**, bool*) in libarrow_flight.a(server.cc.o)
      grpc::internal::CallOpSet<grpc::internal::CallOpRecvMessage<arrow::flight::protocol::HandshakeRequest>, grpc::internal::CallNoOp<2>, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::FillOps(grpc::internal::Call*) in libarrow_flight.a(server.cc.o)
      grpc::internal::CallOpSet<grpc::internal::CallOpRecvMessage<arrow::flight::protocol::HandshakeRequest>, grpc::internal::CallNoOp<2>, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::ContinueFillOpsAfterInterception() in libarrow_flight.a(server.cc.o)
      grpc::internal::CallOpSet<grpc::internal::CallOpRecvMessage<arrow::flight::protocol::HandshakeRequest>, grpc::internal::CallNoOp<2>, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::ContinueFinalizeResultAfterInterception() in libarrow_flight.a(server.cc.o)
      grpc::internal::InterceptorBatchMethodsImpl::Proceed() in libarrow_flight.a(server.cc.o)
      ...
  "grpc::InsecureServerCredentials()", referenced from:
      arrow::flight::FlightServerBase::Init(arrow::flight::FlightServerOptions const&) in libarrow_flight.a(server.cc.o)
  "grpc::MakeChannelArgumentOption(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int)", referenced from:
      grpc::ServerBuilder& grpc::ServerBuilder::AddChannelArgument<int>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int const&) in libarrow_flight.a(server.cc.o)
  "grpc::Status::OK", referenced from:
      arrow::flight::(anonymous namespace)::FlightServiceImpl::ListFlights(grpc::ServerContext*, arrow::flight::protocol::Criteria const*, grpc::ServerWriter<arrow::flight::protocol::FlightInfo>*) in libarrow_flight.a(server.cc.o)
      arrow::flight::(anonymous namespace)::FlightServiceImpl::GetFlightInfo(grpc::ServerContext*, arrow::flight::protocol::FlightDescriptor const*, arrow::flight::protocol::FlightInfo*) in libarrow_flight.a(server.cc.o)
      arrow::flight::(anonymous namespace)::FlightServiceImpl::GetSchema(grpc::ServerContext*, arrow::flight::protocol::FlightDescriptor const*, arrow::flight::protocol::SchemaResult*) in libarrow_flight.a(server.cc.o)
      arrow::flight::(anonymous namespace)::FlightServiceImpl::DoGet(grpc::ServerContext*, arrow::flight::protocol::Ticket const*, grpc::ServerWriter<arrow::flight::protocol::FlightData>*) in libarrow_flight.a(server.cc.o)
      arrow::flight::(anonymous namespace)::FlightServiceImpl::DoAction(grpc::ServerContext*, arrow::flight::protocol::Action const*, grpc::ServerWriter<arrow::flight::protocol::Result>*) in libarrow_flight.a(server.cc.o)
      arrow::flight::(anonymous namespace)::FlightServiceImpl::MakeCallContext(arrow::flight::FlightMethod const&, grpc::ServerContext*, arrow::flight::(anonymous namespace)::GrpcServerCallContext&) in libarrow_flight.a(server.cc.o)
      grpc::Status arrow::flight::(anonymous namespace)::FlightServiceImpl::WriteStream<arrow::flight::FlightInfo, arrow::flight::FlightListing, arrow::flight::protocol::FlightInfo>(arrow::flight::FlightListing*, grpc::ServerWriter<arrow::flight::protocol::FlightInfo>*) in libarrow_flight.a(server.cc.o)
      ...
  "grpc::Status::CANCELLED", referenced from:
      arrow::flight::(anonymous namespace)::FlightServiceImpl::DoAction(grpc::ServerContext*, arrow::flight::protocol::Action const*, grpc::ServerWriter<arrow::flight::protocol::Result>*) in libarrow_flight.a(server.cc.o)
  "grpc::g_glip", referenced from:
      grpc::GrpcLibraryCodegen::GrpcLibraryCodegen(bool) in libarrow_flight.a(protocol_internal.cc.o)
      grpc::GrpcLibraryCodegen::~GrpcLibraryCodegen() in libarrow_flight.a(protocol_internal.cc.o)
  "grpc::ServerContextBase::IsCancelled() const", referenced from:
      arrow::flight::(anonymous namespace)::GrpcServerCallContext::is_cancelled() const in libarrow_flight.a(server.cc.o)
  "grpc::ServerContextBase::peer() const", referenced from:
      arrow::flight::(anonymous namespace)::GrpcServerCallContext::GrpcServerCallContext(grpc::ServerContext*) in libarrow_flight.a(server.cc.o)
  "_grpc_byte_buffer_reader_destroy", referenced from:
      arrow::flight::internal::GrpcBuffer::Wrap(grpc::ByteBuffer*, std::__1::shared_ptr<arrow::Buffer>*) in libarrow_flight.a(serialization_internal.cc.o)
  "_grpc_byte_buffer_reader_init", referenced from:
      arrow::flight::internal::GrpcBuffer::Wrap(grpc::ByteBuffer*, std::__1::shared_ptr<arrow::Buffer>*) in libarrow_flight.a(serialization_internal.cc.o)
  "_grpc_byte_buffer_reader_readall", referenced from:
      arrow::flight::internal::GrpcBuffer::Wrap(grpc::ByteBuffer*, std::__1::shared_ptr<arrow::Buffer>*) in libarrow_flight.a(serialization_internal.cc.o)
  "_grpc_slice_ref", referenced from:
      arrow::flight::internal::GrpcBuffer::GrpcBuffer(grpc_slice, bool) in libarrow_flight.a(serialization_internal.cc.o)
  "_grpc_slice_unref", referenced from:
      arrow::flight::internal::GrpcBuffer::~GrpcBuffer() in libarrow_flight.a(serialization_internal.cc.o)
ld: symbol(s) not found for architecture arm64
```

rok commented:
I've noticed that with the original set of abseil dependencies we were probably getting namespace collisions. That caused some symbols to not be available. Example error message when doing a static build:
```
cmake \
 -GNinja \
 -DCMAKE_INSTALL_PREFIX=$ARROW_HOME \
 -DCMAKE_INSTALL_LIBDIR=lib \
 -DARROW_EXTRA_ERROR_CONTEXT=ON \
 -DARROW_GRPC_USE_SHARED=ON \
 -DARROW_USE_GLOG=ON \
 -DARROW_PARQUET=ON \
 -DARROW_WITH_SNAPPY=ON \
 -DARROW_DATASET=ON \
 -DARROW_CSV=ON \
 -DARROW_JSON=ON \
 -DARROW_PYTHON=ON \
 -DARROW_WITH_LZ4=ON \
 -DARROW_FLIGHT=ON \
 -DARROW_VERBOSE_THIRDPARTY_BUILD=OFF \
 -DCMAKE_BUILD_TYPE="Debug" \
 -DARROW_DEPENDENCY_SOURCE=BUNDLED \
 ..
ninja -j8
```
```
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (any.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(any.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(any.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (any.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(any.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(any.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (arena.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/protobuf_ep-install/lib/libprotobuf.a(arena.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgpr.a(arena.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (descriptor.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(descriptor.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(descriptor.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (duration.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(duration.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(duration.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (duration.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(duration.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(duration.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (empty.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(empty.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(empty.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (empty.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(empty.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(empty.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (http.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(http.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(http.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (http.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(http.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(http.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (insecure_credentials.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(insecure_credentials.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/libgrpcpp.a(insecure_credentials.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (json_util.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(json_util.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/protobuf_ep-install/lib/libprotobuf.a(json_util.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (json_util.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/protobuf_ep-install/lib/libprotobuf.a(json_util.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(json_util.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (metadata.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(metadata.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(metadata.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (metadata.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(metadata.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(metadata.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (parser.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/protobuf_ep-install/lib/libprotobuf.a(parser.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(parser.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (resource.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(resource.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(resource.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (resource.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(resource.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(resource.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (status.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/libgrpcpp.a(status.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/protobuf_ep-install/lib/libprotobuf.a(status.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (status.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(status.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(status.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (status.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(status.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(status.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (struct.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(struct.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(struct.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (struct.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(struct.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(struct.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (time.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgpr.a(time.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/protobuf_ep-install/lib/libprotobuf.a(time.cc.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (timestamp.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(timestamp.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(timestamp.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (timestamp.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(timestamp.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(timestamp.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (wrappers.upb.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(wrappers.upb.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(wrappers.upb.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (wrappers.upbdefs.c.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libupb.a(wrappers.upbdefs.c.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(wrappers.upbdefs.c.o) due to use of basename, truncation and blank padding
/Library/Developer/CommandLineTools/usr/bin/libtool: warning same member name (xds_credentials.cc.o) in output file used for input files: /Users/rok/Documents/repos/ARROW-14708/cpp/build/libgrpcpp.a(xds_credentials.cc.o) and: /Users/rok/Documents/repos/ARROW-14708/cpp/build/grpc_ep-install/lib/libgrpc.a(xds_credentials.cc.o) due to use of basename, truncation and blank padding
```

This lead me to removing some abseil packages from flight, but I can't solve the missing gRPC probably because it's bundled differently.

lidavidm commented:
Looking at CI, it seems the set of Abseil libraries we need might be different per-platform even.

Also, I get failures with the repro. The link command is missing the bundled dependencies, and the order of libraries still seems off. (Can't paste the log, it's too long.)

```
[1/1] Linking CXX executable arrow_example
FAILED: arrow_example 
: && /usr/bin/c++ -O3 -DNDEBUG  CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example  /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow.a  /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_flight.a  -pthread && :
```

Even if I change the order of linking and add the bundled dependencies, I still have gRPC symbols missing. And indeed, the dependencies archive seems to be missing gRPC.

```
CREATE /home/lidavidm/Code/upstream/rok-arrow-14708/build/debug/libarrow_bundled_dependencies.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/snappy_ep/src/snappy_ep-install/lib/libsnappy.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/glog_ep-prefix/src/glog_ep/lib/libglogd.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/gflags_ep-prefix/src/gflags_ep/lib/libgflags_debug.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/thrift_ep-install/lib/libthriftd.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/protobuf_ep-install/lib/libprotobuf.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/jemalloc_ep-prefix/src/jemalloc_ep/dist//lib/libjemalloc_pic.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/zlib_ep/src/zlib_ep-install/lib/libz.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/lz4_ep-prefix/src/lz4_ep/lib/liblz4.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/utf8proc_ep-install/lib/libutf8proc.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/cares_ep-install/lib/libcares.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_base.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_cord.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_synchronization.a
SAVE
END
```

rok commented:
@github-actions crossbow submit example-cpp-minimal-build-static*

rok commented:
@kou I think this is ready for review could you please take a look?

Please note cpp/build-support/list-absl-libraries.sh seems a bit unnecessary now. We should probably remove it.

kou commented:
Could you confirm test failures?

kou commented:
It seems that they are related because they are failed by link error:

https://github.com/apache/arrow/runs/5386081478?check_suite_focus=true

```text
Error: package or namespace load failed for ‘arrow’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new/arrow/libs/arrow.so':
  /arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new/arrow/libs/arrow.so: undefined symbol: _ZN3re23RE2D1Ev
```

https://github.com/apache/arrow/runs/5386081500?check_suite_focus=true
  
```text
Error: package or namespace load failed for ‘arrow’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new/arrow/libs/arrow.so':
  /arrow/r/check/arrow.Rcheck/00LOCK-arrow/00new/arrow/libs/arrow.so: undefined symbol: _ZN3re23RE2D1Ev
```  

https://github.com/ursacomputing/crossbow/runs/5386355726?check_suite_focus=true

```text
[100%] Linking CXX executable arrow_example
/usr/bin/ld: /usr/local/lib/libarrow.a(compression_brotli.cc.o): in function `arrow::util::internal::(anonymous namespace)::BrotliDecompressor::~BrotliDecompressor()':
compression_brotli.cc:(.text+0x13c): undefined reference to `BrotliDecoderDestroyInstance'
/usr/bin/ld: /usr/local/lib/libarrow.a(compression_brotli.cc.o): in function `std::_Sp_counted_ptr_inplace<arrow::util::internal::(anonymous namespace)::BrotliDecompressor, std::allocator<arrow::util::internal::(anonymous namespace)::BrotliDecompressor>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()':
...
```

lidavidm commented:
This still gives me a lot of errors with the repro:

```
> ninja                                                                                                            (base) 
[1/1] Linking CXX executable arrow_example
FAILED: arrow_example 
: && /usr/bin/c++ -O3 -DNDEBUG  CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example  /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow.a  /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_flight.a  /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_bundled_dependencies.a  -pthread && :
/home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_flight.a(server.cc.o): In function `arrow::flight::(anonymous namespace)::DoExchangeMessageWriter::DoExchangeMessageWriter(grpc::ServerReaderWriter<arrow::flight::protocol::FlightData, arrow::flight::protocol::FlightData>*)':
/home/lidavidm/Code/upstream/rok-arrow-14708/cpp/src/arrow/flight/server.cc:301: undefined reference to `arrow::ipc::IpcWriteOptions::Defaults()'
/home/lidavidm/Code/upstream/rok-arrow-14708/cpp/src/arrow/flight/server.cc:301: undefined reference to `arrow::ipc::DictionaryFieldMapper::DictionaryFieldMapper()'
/home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_flight.a(server.cc.o): In function `arrow::flight::(anonymous namespace)::DoExchangeMessageWriter::Begin(std::shared_ptr<arrow::Schema> const&, arrow::ipc::IpcWriteOptions const&)':
...
```

The link order is still wrong - libarrow_flight.a needs to come before libarrow.a. After that, it seems gRPC isn't getting bundled. Indeed, the .ar file is missing gRPC:

```
CREATE /home/lidavidm/Code/upstream/rok-arrow-14708/build/debug/libarrow_bundled_dependencies.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/snappy_ep/src/snappy_ep-install/lib/libsnappy.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/glog_ep-prefix/src/glog_ep/lib/libglogd.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/gflags_ep-prefix/src/gflags_ep/lib/libgflags_debug.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/thrift_ep-install/lib/libthriftd.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/protobuf_ep-install/lib/libprotobuf.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/jemalloc_ep-prefix/src/jemalloc_ep/dist//lib/libjemalloc_pic.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/zlib_ep/src/zlib_ep-install/lib/libz.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/lz4_ep-prefix/src/lz4_ep/lib/liblz4.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/re2_ep-install/lib/libre2.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/utf8proc_ep-install/lib/libutf8proc.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/cares_ep-install/lib/libcares.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_base.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_cord.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_status.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_statusor.a
ADDLIB /home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/lib/libabsl_synchronization.a
SAVE
END
```

lidavidm commented:
aaaaand when I try to fix it: https://github.com/google/shaderc/issues/473#issuecomment-409754031

…

lidavidm commented:
So, I can manually fix that by copying libgrpc++.a to libgrpcpp.a, then running ar myself. So far, these changes have been made:

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 84c27c254..220a5ded0 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -3647,6 +3647,8 @@ macro(build_grpc)
   add_custom_target(grpc_copy_grpc++ ALL DEPENDS "${GRPC_STATIC_LIBRARY_GRPCPP_FOR_AR}")
   add_dependencies(gRPC::grpcpp_for_bundling grpc_copy_grpc++)
 
+  list(APPEND ARROW_BUNDLED_STATIC_LIBS gRPC::upb gRPC::gpr gRPC::address_sorting
+         gRPC::grpc++ gRPC::grpc)
   list(APPEND ARROW_BUNDLED_STATIC_LIBS ${GRPC_GPR_ABSL_LIBRARIES})
 endmacro()
```

and I'm running

```
/usr/bin/c++ -O3 -DNDEBUG  CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_flight.a /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow.a /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_bundled_dependencies.a  -pthread
``` 

That fails with more missing symbols. A lot look like OpenSSL, so that's another issue here: something isn't properly declaring the dependency on OpenSSL.

```
/usr/bin/c++ -O3 -DNDEBUG  CMakeFiles/arrow_example.dir/example.cc.o -o arrow_example /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_flight.a /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow.a /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_bundled_dependencies.a -lssl -lcrypto  -pthread
```

This still fails, now we're missing more abseil symbols:

```
/home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_bundled_dependencies.a(completion_queue.cc.o): In function `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > absl::lts_20210324::StrFormat<void*>(absl::lts_20210324::str_format_internal::FormatSpecTemplate<(ArgumentToConv<void*>)()> const&, void* const&)':
/home/lidavidm/Code/upstream/rok-arrow-14708/build/absl_ep-install/include/absl/strings/str_format.h:342: undefined reference to `absl::lts_20210324::str_format_internal::FormatPack[abi:cxx11](absl::lts_20210324::str_format_internal::UntypedFormatSpecImpl, absl::lts_20210324::Span<absl::lts_20210324::str_format_internal::FormatArgImpl const>)'
/home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/libarrow_bundled_dependencies.a(completion_queue.cc.o): In function `void absl::lts_20210324::str_format_internal::FormatArgImpl::Init<absl::lts_20210324::str_format_internal::VoidPtr>(absl::lts_20210324::str_format_internal::VoidPtr const&)':
...
```

@rok does the repro work for you?

lidavidm commented:
Aren't a lot of these issues the same ones we had initially? Especially the missing OpenSSL and the wrong order of the link parameters.

rok commented:
I started working on this on mac and was getting namespace collision and therefore removed some abseils. I was also [hitting this](https://github.com/abseil/abseil-cpp/issues/283) and couldn't get around it so I decided to focus on linux.

rok commented:
@lidavidm Thanks for the suggestions, let me try them together with reintroducing those abseils.

rok commented:
> @rok does the repro work for you?

No, static examples do not build yet.

lidavidm commented:
Ah, ok, understood. Hmm, we should probably see about setting this case up in Crossbow once we get it fixed.

rok commented:
> Ah, ok, understood. Hmm, we should probably see about setting this case up in Crossbow once we get it fixed.

That would be a good idea. We should make sure we check the mac issues once we're done here, but I would prefer that in a separate scope.

rok commented:
@lidavidm this now runs for me on linux (conda env on ubuntu) if I replace `cpp/examples/minimal_build/example.cc` and `cpp/examples/minimal_build/CMakeLists.txt` with [repro code](https://issues.apache.org/jira/browse/ARROW-14708):
```
cmake .. \
  -DCMAKE_BUILD_TYPE=Debug \
  -GNinja \
  -DCMAKE_INSTALL_PREFIX=$ARROW_HOME \
  -DCMAKE_INSTALL_LIBDIR=lib \
  -DARROW_USE_GLOG=ON \
  -DARROW_PARQUET=ON \
  -DARROW_WITH_SNAPPY=ON \
  -DARROW_DATASET=ON \
  -DARROW_CSV=ON \
  -DARROW_JSON=ON \
  -DARROW_PYTHON=ON \
  -DARROW_WITH_LZ4=ON \
  -DARROW_FLIGHT=ON \
  -DARROW_DEPENDENCY_SOURCE=BUNDLED

ninja -j8
```
I'll check on mac shortly but I suspect it won't work there/

rok commented:
Seems like there's an unrelated issue with [AWSSDKConfig.cmake](https://github.com/apache/arrow/runs/5411952297?check_suite_focus=true#step:6:1674).

rok commented:
On mac:
```
Undefined symbols for architecture arm64:
  "_CFRelease", referenced from:
      absl::lts_20210324::time_internal::cctz::local_time_zone() in libabsl_time_zone.a(time_zone_lookup.cc.o)
  "_CFStringGetCString", referenced from:
      absl::lts_20210324::time_internal::cctz::local_time_zone() in libabsl_time_zone.a(time_zone_lookup.cc.o)
  "_CFStringGetLength", referenced from:
      absl::lts_20210324::time_internal::cctz::local_time_zone() in libabsl_time_zone.a(time_zone_lookup.cc.o)
  "_CFStringGetMaximumSizeForEncoding", referenced from:
      absl::lts_20210324::time_internal::cctz::local_time_zone() in libabsl_time_zone.a(time_zone_lookup.cc.o)
  "_CFTimeZoneCopyDefault", referenced from:
      absl::lts_20210324::time_internal::cctz::local_time_zone() in libabsl_time_zone.a(time_zone_lookup.cc.o)
  "_CFTimeZoneGetName", referenced from:
      absl::lts_20210324::time_internal::cctz::local_time_zone() in libabsl_time_zone.a(time_zone_lookup.cc.o)
```
Which looks like [this cctz issue](https://github.com/abseil/abseil-cpp/issues/283).

lidavidm commented:
It's still failing for me.

lidavidm commented on cpp/src/arrow/CMakeLists.txt:
```diff
@@ -534,7 +534,8 @@ else()
 endif()
 
 if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
-  string(APPEND ARROW_PC_LIBS_PRIVATE " -larrow_bundled_dependencies")
+else()
```
`if (NOT ...)`?

rok replied:
Done

rok commented:
@lidavidm I've reintroduced all abseils that were there before and hopefully changed the order of imports. Could you please check again? Strangely this keeps on building for me here.

lidavidm commented:
I wonder if it's because one of us is enabling C++17? Apparently that changes the available symbols: https://github.com/conda-forge/abseil-cpp-feedstock/pull/25#issuecomment-1058682145

lidavidm commented:
I'll try with the docker container now

lidavidm commented:
Wait, what changes exactly are you making to the minimal build? (Can you give a diff?) The example there doesn't actually link to or use Flight by default, so it might not complain.

rok commented:
@lidavidm Ah sorry, I was building the repro with `-DARROW_BUILD_EXAMPLES=ON` and wasn't getting issues. Building with:
```
cmake . -DCMAKE_MODULE_PATH=/home/rok/Documents/repos/arrow/cpp/cmake_modules/ -DCMAKE_PREFIX_PATH=/home/rok/miniconda3/envs/pyarrow-dev/lib/
```
Surfaces the symbol issues again. I'll fix it and get back.

rok commented:
@lidavidm this now builds the repro on my laptop with a small adjustment to CMakeLists.txt. Namely last line should be:
`target_link_libraries(arrow_example PRIVATE arrow_flight_static arrow_static arrow_bundled_dependencies ssl crypto rt Threads::Threads)`.
Could you please check again on your end?

lidavidm commented:
For some reason it now refuses to link:

```
> cmake ../ -GNinja -DCMAKE_PREFIX_PATH=../install/lib/cmake/arrow
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.1.1")  
CMake Error at /home/lidavidm/Code/upstream/rok-arrow-14708/install/lib/cmake/arrow/ArrowConfig.cmake:101 (set_target_properties):
  set_target_properties called with incorrect number of arguments.
Call Stack (most recent call first):
  CMakeLists.txt:5 (find_package)


-- Arrow version: 8.0.0-SNAPSHOT
-- Arrow SO version: 800.0.0
-- Configuring incomplete, errors occurred!
See also "/home/lidavidm/Code/upstream/rok-arrow-14708/repro/build/CMakeFiles/CMakeOutput.log".
See also "/home/lidavidm/Code/upstream/rok-arrow-14708/repro/build/CMakeFiles/CMakeError.log".
```

The problem is here:
```
101       set_target_properties(
102         arrow_bundled_dependencies
103         PROPERTIES
104           IMPORTED_LOCATION
105           "${arrow_lib_dir}/${CMAKE_STATIC_LIBRARY_PREFIX}arrow_bundled_dependencies${CMAKE_STATIC_LIBRARY_SUFFIX}"
106           INTERFACE_LINK_LIBRARIES
107           ${ARROW_STATIC_INSTALL_INTERFACE_LIBS}
108       )
```

`ARROW_STATIC_INSTALL_INTERFACE_LIBS` is blank:

```
 65 set(ARROW_STATIC_INSTALL_INTERFACE_LIBS "")
```

lidavidm commented:
This fixes it:

```diff
diff --git a/cpp/src/arrow/ArrowConfig.cmake.in b/cpp/src/arrow/ArrowConfig.cmake.in
index e8d79c089..1f4ae96b1 100644
--- a/cpp/src/arrow/ArrowConfig.cmake.in
+++ b/cpp/src/arrow/ArrowConfig.cmake.in
@@ -80,7 +80,7 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
           IMPORTED_LOCATION
           "${arrow_lib_dir}/${CMAKE_STATIC_LIBRARY_PREFIX}arrow_bundled_dependencies${CMAKE_STATIC_LIBRARY_SUFFIX}"
           INTERFACE_LINK_LIBRARIES
-          ${ARROW_STATIC_INSTALL_INTERFACE_LIBS}
+          "${ARROW_STATIC_INSTALL_INTERFACE_LIBS}"
       )
       get_property(arrow_static_interface_link_libraries
                    TARGET arrow_static
```

Also if we change this, we don't need to list ssl, crypto, etc. in the downstream package:

```diff
diff --git a/cpp/src/arrow/CMakeLists.txt b/cpp/src/arrow/CMakeLists.txt
index 403a4e39a..b206ba894 100644
--- a/cpp/src/arrow/CMakeLists.txt
+++ b/cpp/src/arrow/CMakeLists.txt
@@ -533,7 +533,7 @@ else()
   set(ARROW_BUILD_BUNDLED_DEPENDENCIES FALSE)
 endif()
 
-if(NOT ARROW_BUILD_BUNDLED_DEPENDENCIES)
+if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
   set(ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
 endif()
 # Need -latomic on Raspbian.
```

But it finally works!

lidavidm commented on cpp/cmake_modules/BuildUtils.cmake:
```diff
@@ -240,7 +240,9 @@ function(ADD_ARROW_LIB LIB_NAME)
-      OUTPUT_PATH)
+      OUTPUT_PATH
+      SHARED_INSTALL_EXTRA_TARGETS
+      STATIC_INSTALL_EXTRA_TARGETS)
```
Do we actually use any of the changes here anymore?

rok replied:
Removed.

lidavidm replied:
I'm still confused here, now the arguments were added back but we don't use them?

lidavidm commented on cpp/src/arrow/flight/CMakeLists.txt:
```diff
@@ -197,7 +197,9 @@ add_arrow_lib(arrow_flight
               arrow_static
-              ${ARROW_FLIGHT_LINK_LIBS})
+              ${ARROW_FLIGHT_LINK_LIBS}
+              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL}
+              arrow_static)
```
Hmm, do we need to list arrow_static twice? And it seems odd that we have to re-list all the libraries here…why doesn't listing arrow_static as a dependency give us these dependencies transitively?

rok replied:
Indeed. Removed.

rok commented:
@lidavidm that's great! Thanks for the review I'll address the comments and get back.

Question: should we add a test for the static build?

lidavidm commented on cpp/cmake_modules/BuildUtils.cmake:
```diff
@@ -406,7 +406,8 @@ function(ADD_ARROW_LIB LIB_NAME)
     endif()
 
-    install(TARGETS ${LIB_NAME}_shared ${INSTALL_IS_OPTIONAL}
+    install(TARGETS ${LIB_NAME}_shared ${ARG_SHARED_INSTALL_EXTRA_TARGETS}
+                    ${INSTALL_IS_OPTIONAL}
```
This change and the one below are also unnecessary right?

rok replied:
Removed.

rok commented on cpp/src/arrow/flight/CMakeLists.txt:
```diff
@@ -197,7 +197,8 @@ add_arrow_lib(arrow_flight
               arrow_static
-              ${ARROW_FLIGHT_LINK_LIBS})
+              ${ARROW_FLIGHT_LINK_LIBS}
+              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL})
```
Removed.

rok commented:
@lidavidm Implemented your suggestions and added a workaround for the [abseil issue](https://github.com/abseil/abseil-cpp/issues/283) on macOS.

lidavidm commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -3403,15 +3403,32 @@ macro(build_absl_once)
+                            absl::time_zone
+                            ${CoreFoundation})
+    else()
+      find_library(CoreFoundation CoreFoundation)
```
This won't work on non-Apple platforms will it?

lidavidm replied:
Do we need this branch at all? Presumably absl::time's deps are correct on non-Apple platforms and we don't need to overwrite them?

rok replied:
Sorry, that shouldn't be there.

rok replied:
I'm not sure about absl::time's deps on non-Apple so leaving this as is.

lidavidm commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -3403,15 +3403,32 @@ macro(build_absl_once)
+      # This is due to upstream absl::cctz issue
+      # https://github.com/abseil/abseil-cpp/issues/283
+      find_library(CoreFoundation CoreFoundation)
+      set_property(TARGET absl::time
```
Can we append to the property instead of having to list out all the dependencies?

rok replied:
Switched to append.

lidavidm commented on cpp/examples/arrow/CMakeLists.txt:
```diff
@@ -42,8 +42,12 @@ if(ARROW_FLIGHT)
-    set(GRPC_REFLECTION_LINK_LIBS -Wl,--whole-archive gRPC::grpc++_reflection
-                                  -Wl,--no-whole-archive)
+    if(APPLE)
+      set(GRPC_REFLECTION_LINK_LIBS -Wl,-force_load gRPC::grpc++_reflection -Wl)
```
Shouldn't this end with `-Wl,--no-force-load` or something?

Related, does it take `-force-load` or `--force-load`?

rok replied:
So `all_load`/`noall_load` currently won't work due to symbol duplication so `force-load` is the only option and there appears to be no `noforce-load` option.
More [here](https://github.com/premake/premake-core/issues/1170).

lidavidm commented on cpp/src/arrow/CMakeLists.txt:
```diff
@@ -533,8 +533,8 @@ else()
 
-if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
-  string(APPEND ARROW_PC_LIBS_PRIVATE " -larrow_bundled_dependencies")
+if(NOT ARROW_BUILD_BUNDLED_DEPENDENCIES)
```
Is this back to `NOT`? I feel like it flip-flops on every commit…was there a rebase error or something? It might be easier to squash current commits if needed

rok replied:
Ugh, sorry. I was switching back and forth between linux and mac to fix the mac issue and probably caused this. I'll squash later today.

rok commented:
@lidavidm thanks for your patience. I've pushed another change with your suggestions.

lidavidm commented on cpp/cmake_modules/BuildUtils.cmake:
```diff
@@ -239,8 +239,7 @@ function(ADD_ARROW_LIB LIB_NAME)
       PRIVATE_INCLUDES
       DEPENDENCIES
       SHARED_INSTALL_INTERFACE_LIBS
-      STATIC_INSTALL_INTERFACE_LIBS
-      OUTPUT_PATH)
```
Hmm, did you mean to remove this parameter?

rok replied:
Restored.

lidavidm commented on cpp/src/arrow/ArrowConfig.cmake.in:
```diff
@@ -79,11 +80,12 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
-        )
-
+          INTERFACE_LINK_LIBRARIES
+          ${ARROW_STATIC_INSTALL_INTERFACE_LIBS}
```
This needs to be quoted.

rok replied:
Done.

lidavidm reviewed (approved):
Thanks! Things work for me now.

rok commented:
Sorry for the rebasing mess @lidavidm ! :)

kou commented on cpp/src/arrow/CMakeLists.txt:
```diff
@@ -534,7 +534,7 @@ else()
 endif()
 
 if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
-  string(APPEND ARROW_PC_LIBS_PRIVATE " -larrow_bundled_dependencies")
```
Why is this removed?

rok replied:
Returned.

kou commented on cpp/src/arrow/CMakeLists.txt:
```diff
@@ -563,10 +563,10 @@ add_arrow_lib(arrow
               ${ARROW_SHARED_INSTALL_INTERFACE_LIBS}
-              STATIC_INSTALL_INTERFACE_LIBS
-              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
+              STATIC_INSTALL_INTERFACE_LIBS)
```
Can we remove `STATIC_INSTALL_INTERFACE_LIBS`?

rok replied:
It seems so. Changed.

kou commented on cpp/src/arrow/CMakeLists.txt:
```diff
@@ -534,7 +534,7 @@ else()
 
 if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
-  string(APPEND ARROW_PC_LIBS_PRIVATE " -larrow_bundled_dependencies")
+  set(ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
```
Why do we need `_REAL` alias?
Is there any problem if we use `ARROW_STATIC_INSTALL_INTERFACE_LIBS` directly?

rok commented:
@kou thank you for the review! I've pushed suggested changes.
